media/src/androidLibMain/kotlin/DefaultMediaManager.kt

```diff
@@ -0,0 +1,202 @@
+    }
+
+    private val mediaPlayer = AndroidMediaPlayer()
+    private val volume = MutableStateFlow(1.0f)
+    override val currentVolume: Flow<Float> = volume.asSharedFlow()
+    override suspend fun updateVolume(volume: Float) {
+        mediaPlayer.setVolume(volume, volume)
+        this.volume.value = volume
+    }
+
+    init {
+        mediaPlayer.setOnSeekCompleteListener {
+            handleSeekCompleted(true)
+        }
+        mediaPlayer.setOnCompletionListener { handleCompleted() }
+        mediaPlayer.setOnErrorListener { _, _, extra ->
+            val error = when (extra) {
+                AndroidMediaPlayer.MEDIA_ERROR_IO -> PlaybackError.IO
+                AndroidMediaPlayer.MEDIA_ERROR_MALFORMED -> PlaybackError.MalformedMediaSource
+                AndroidMediaPlayer.MEDIA_ERROR_UNSUPPORTED -> PlaybackError.Unsupported
+                AndroidMediaPlayer.MEDIA_ERROR_NOT_VALID_FOR_PROGRESSIVE_PLAYBACK -> PlaybackError.Unsupported
+                AndroidMediaPlayer.MEDIA_ERROR_TIMED_OUT -> PlaybackError.TimedOut
+                else -> PlaybackError.Unknown
+            }
+            handleError(error)
+            false
+        }
+    }
+
+    override fun cleanUp() {
+        mediaPlayer.release()
+    }
+
+    override fun handleCreatePlayableMedia(source: MediaSource): PlayableMedia? = try {
+        when (source) {
+            is MediaSource.Asset -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                mediaPlayer.setDataSource(source.descriptor)
+            } else {
+                mediaPlayer.setDataSource(source.descriptor.fileDescriptor)
+            }
+            is MediaSource.File -> mediaPlayer.setDataSource(source.descriptor)
+            is MediaSource.Url -> mediaPlayer.setDataSource(source.url.toExternalForm())
+            is MediaSource.Content -> if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                mediaPlayer.setDataSource(source.context, source.uri, source.headers, source.cookies)
+            } else {
+                mediaPlayer.setDataSource(source.context, source.uri, source.headers)
+            }
+        }
+        DefaultPlayableMedia(source, mediaPlayer)
+    } catch (e: Throwable) {
+        when (e) {
+            is IllegalStateException -> null
+            is IOException -> null
+            is IllegalArgumentException -> null
+            is SecurityException -> null
+            else -> throw e
+        }
+    }
+
+    override fun initialize(playableMedia: PlayableMedia) {
+        mediaPlayer.setOnPreparedListener {
+            mediaPlayer.setOnPreparedListener(null)
+            handlePrepared(DefaultPlayableMedia(playableMedia.source, it))
+        }
+        mediaPlayer.prepareAsync()
+    }
+
+    override suspend fun renderVideoOnSurface(surface: MediaSurface?) {
+        try {
+            mediaPlayer.setDisplay(surface?.holder)
+        } catch (e: IllegalStateException) {
+            //
```

Comment:
Yeah its mostly here because cleanup sets it to null, but that actually throws this exception if the Android Media Player has already been ended. So I just catch it for this rare edgecase. Other cases *should* be accounted for by the state machine